Authentication - Tyk and Keycloak

amanjeev opened the pull request:
This is to be merged before https://github.com/CanDIG/CanDIGv2/pull/91

- Adds Tyk and Keycloak deployment 
- Moves Tyk and Keycloak etc. to top level citizens
- Uses `Makefile.authx` to logically group authentication and authorization into a unit
- Does NOT automatically add authx to provision.sh

amanjeev committed:
update (conda): settings that allow for conda env setup without intervention

DIG-633

amanjeev committed:
add (vagrant): libvirt section to launch using stuff like QEMU

DIG-633

amanjeev committed:
refactor (authx): reorganizing structure; WIP; DIG-633

amanjeev committed:
chore (README): spelling

amanjeev committed:
chore (gitignore): add .idea directory

amanjeev committed:
feature (vagrant): add IP address option to Vagrantfile

amanjeev committed:
feature (authx): add keycloak to the setup launch

amanjeev committed:
feature (authx): minor formatting for keycloak scripts

DIG-633

amanjeev committed:
feature (authx): minor formatting for keycloak scripts

DIG-633

amanjeev committed:
feature (authx): WIP tyk service, simplifying setup

DIG-633

amanjeev committed:
feature (authx): WIP tyk service, simplifying setup

moving tmp inside lib/tyk alleviates this pain for now but
this is not a good solution as it breaks the repo convention.

DIG-633

amanjeev committed:
refactor (authx): KEYCLOAK_SERVICE* to KEYCLOAK*

DIG-633

amanjeev committed:
feature (authx): CHECKPOINT in case of fire

DIG-633

amanjeev committed:
feature (authx): add + as exclusion in makefile secret generator

DIG-633

amanjeev committed:
feature (authx): fix tyk confs url

DIG-633

amanjeev committed:
feature (authx): remove candig-server from authx makefile because it is already launched; add image removal in cleanup; DIG-633

amanjeev committed:
feature (authx): refactor variables in keycloak script to remove global and rename locals; DIG-633

amanjeev committed:
feature (authx): formatting; DIG-633

amanjeev committed:
feature (authx): fix tyk redirect uri instead of candig server in keycloak client redirect uri settings; DIG-633

amanjeev committed:
feature (authx): add security TODO warning; DIG-633

amanjeev committed:
feature (authx): CHECKPOINT in case of fire, working on tyk;

DIG-633

amanjeev committed:
feature (authx): Tyk api redirect works

DIG-633

amanjeev committed:
docs (authx): document steps, and a todo

DIG-633

amanjeev committed:
refactor (authx): renames TEMP_KEYCLOAK.. to KEYCLOACL...PROD because thats the purpose of that URL

adds the variable to environment

DIG-633

amanjeev committed:
feature (authx): analytics for tyk

DIG-633

amanjeev committed:
feature (authx): remove check for local idp for now

DIG-633

amanjeev committed:
feature (authx): add warning comments

DIG-633

amanjeev committed:
feature (authx): add directory cleanup for tyk tmp

DIG-633

amanjeev committed:
feature (authx): add directory cleanup for tyk tmp

DIG-633

amanjeev committed:
docs (authx): adds new api section because we need to convey that

right now the tyk setup is adhoc at best, it deploys fine with single
api (candig) but it is not enough. this section documents how to achieve
this in a hacky way. it is rather sad but it is also need of the hour.

DIG-633

amanjeev committed:
docs (authx): steps to add new api

amanjeev committed:
feature (authx): resolve conflicts; DIG-633

amanjeev committed:
feature (authx): resolve conflicts because I clearly cannot read; DIG-633

daisieh commented on lib/tyk/tyk_setup.sh:
```diff
@@ -0,0 +1,68 @@
+
+mkdir -p $CONFIG_DIR $CONFIG_DIR/apps $CONFIG_DIR/policies $CONFIG_DIR/middleware
+
+echo "Working on tyk.conf"
```
Would you consider adding ` | tee -a $LOGFILE` to the echo commands (as in provision.sh), so that they'll end up in the progress file?

daisieh commented on docs/authx-setup.md:
```diff
@@ -0,0 +1,53 @@
+  the key `access_rights`
+- Add the new line to copy the file to the image in the `lib/tyk/Dockerfile`
+- Add the new line to `envsubst` in `lib/tyk/tyk_setup.sh` (see section `# Extra APIs can be added here`)
+- Redeploy the container OR use Tyk API (TODO: ask Jimmy about this)
```
You should be able to call the POST /tyk/apis endpoint described [here](https://www.tyk.io/docs/tyk-gateway-api/)...just call it with the contents of the json file as the body and it'll hot-load it.

amanjeev replied:
Thank you for this! They do not list version 2.x in the doc you provided.  I think to use this feature I first have to upgrade Tyk to 3.x and then do this.  Would you be ok if we created a jira ticket for the improvement(s) you list in this comment and tackle them later?

amanjeev replied:
@daisieh I am going to attempt to bump tyk version and if I succeed I will report here and update this PR.  Then I will create a new ticket for improvement you suggest.  Does that sounds ok to you?

daisieh replied:
I see that you updated the tyk version...does that mean that we can now use the /tyk/apis endpoint? If so, can you update this doc?

amanjeev replied:
@daisieh yes version is updated but I have no idea about the api endpoint until I try it out.  What I have tried out is file based stuff so I am not comfortable changing the docs yet.

daisieh reviewed (changes requested):
One of my comments is more annoying than the other: basically, giant find-and-replace all of the echos with also piping the echo through `tee` to get it into the progress log file (` | tee -a $LOGFILE`).

Otherwise, as far as I can tell from my understanding of these things, it looks good.

amanjeev committed:
feature (authx): add `tee` to logfile

Suggestion at https://github.com/CanDIG/CanDIGv2/pull/99#discussion_r736020301

DIG-633

amanjeev commented:
I learned that I have failing health checks.  I am fixing those and will push changes soon.

amanjeev committed:
feature (authx): fix failing incorrect health checks for containers

DIG-633

amanjeev commented:
> I learned that I have failing health checks. I am fixing those and will push changes soon.

This is done

amanjeev committed:
chore (authx): bumps up version of tyk and redis

DIG-633

amanjeev commented:
Also, bumped up the tyk and redis version.

shaikh-rashid commented on Makefile:
```diff
@@ -27,13 +28,13 @@ all:
 #<<<
 .PHONY: mkdir
 mkdir:
+	mkdir -p $(DIR)/bin
```
repeat line

shaikh-rashid commented on Makefile.authx:
```diff
@@ -0,0 +1,55 @@
+		$(DIR)/lib/keycloak/docker-compose.yml | docker-compose -f - down
+
+	cat $(DIR)/lib/compose/docker-compose.yml $(DIR)/lib/logging/$(DOCKER_LOG_DRIVER)/docker-compose.yml \
+    	$(DIR)/lib/tyk/docker-compose.yml | docker-compose -f - down
```
fix indentation

shaikh-rashid commented on Makefile.authx:
```diff
@@ -0,0 +1,55 @@
+	docker rmi candig_tyk:latest --force
+
+	# - clean tmp dir inside lib/tyk
+	rm -r $(DIR)/lib/tyk/tmp
```
does tyk tmp need to exist inside /lib/tyk ? can it be in /tmp/data/tyk ?

shaikh-rashid replied:
nvm i recall why this needed to be done

amanjeev replied:
yea I just gave up early tbh

shaikh-rashid commented on Makefile.authx:
```diff
@@ -0,0 +1,55 @@
+
+
+#>>>
+# authx, common settings
```
missing newline

shaikh-rashid commented on etc/env/example.env:
```diff
@@ -9,14 +9,6 @@ CANDIG_AUTH_MODULES=keycloak tyk opa vault arbiter
 CANDIG_DOMAIN=docker.localhost
 CANDIG_AUTH_DOMAIN=auth.docker.localhost
 
-# candig server
```
why deprecate?

amanjeev replied:
not deprecated, this was a duplicate and is mentioned again below on this line https://github.com/CanDIG/CanDIGv2/blob/amanjeev/dig-633-authentication/etc/env/example.env#L193-L198

shaikh-rashid commented on etc/env/example.env:
```diff
@@ -226,31 +219,36 @@ KEYCLOAK_REALM=candig
 TYK_CANDIG_API_ID=21
 TYK_CANDIG_API_NAME=CanDIG
 TYK_CANDIG_API_SLUG=candig
+TYK_CANDIG_API_TARGET=http://${CANDIG_DOMAIN}:${CANDIG_SERVER_PORT}
```
you are still using candig-server env vars

amanjeev replied:
yes they are define here https://github.com/CanDIG/CanDIGv2/blob/amanjeev/dig-633-authentication/etc/env/example.env#L193-L198

shaikh-rashid commented on etc/env/example.env:
```diff
@@ -274,11 +272,6 @@ OPA_URL=http://opa:${OPA_PORT}
 
 CANDIG_AUTHZ_SERVICE_PORT=8182
 
-# same as former candig-server port - used to relay to new c-s port
```
just want to confirm that we are deprecating arbiter code

amanjeev replied:
yes, we are removing arbiter.  

arbiter was never and is of no use as proposed.  arbiter's purpose was proposed to be the policy enforcer but that does not work unless arbiter sits in the middle of the application and the database and at the same time speaks the said database's wire protocol to munge sql request to the database.  which is not something that we are doing or will do (although this is a good research problem, right @ljdursi ? ).

amanjeev replied:
this reminds me that i forgot to remove arbiter from the auth module list.  thanks for mentioning.

Done!

shaikh-rashid reviewed (changes requested):
Excellent work. Minor changes and clarifications requested. Thx

amanjeev committed:
fix (authx): remove repeat line

https://github.com/CanDIG/CanDIGv2/pull/99#discussion_r746730247

DIG-633

amanjeev committed:
fix (authx): indentation should be tabs, not 4 spaces in Make

https://github.com/CanDIG/CanDIGv2/pull/99#discussion_r746732718

DIG-633

amanjeev committed:
fix (authx): missing new line

https://github.com/CanDIG/CanDIGv2/pull/99#discussion_r746745183

DIG-633

amanjeev committed:
fix (authx): remove arbiter

https://github.com/CanDIG/CanDIGv2/pull/99#discussion_r746992537

DIG-633

daisieh reviewed (approved):
Looks good, thanks!
